recipes/dev/grpo_full_finetune_distributed.py

```diff
@@ -0,0 +1,1289 @@
+            # in case shuffle is True
+            self._sampler.set_epoch(curr_epoch)
+
+            pbar = tqdm(total=self._steps_per_epoch, disable=not self._is_rank_zero)
+            for idx, batch in enumerate(self._dataloader):
+                if (
+                    self.max_steps_per_epoch is not None
+                    and (idx // self._gradient_accumulation_steps)
+                    == self.max_steps_per_epoch
+                ):
+                    break
+
+                # Start tracking CUDA memory for active steps for just the first epoch
+                if (
+                    self._is_rank_zero
+                    and curr_epoch == 0
+                    and self.profiler_profile_memory
+                    and idx == self.profiler_wait_steps + self.profiler_warmup_steps
+                ):
+                    torch.cuda.memory._record_memory_history()
+
+                tokens = batch["tokens"]  # type: ignore
+                answers = batch["answers"]  # type: ignore
+                tokens = tokens.to(self._device)  # [B, P]
+
+                _, context_length = tokens.shape
+
+                trajectory = self.generate_trajectory_batched(tokens, answers)
+
+                # Calculate the number of unmasked tokens in the current batch
+                # and increment the total number of tokens seen in the step
+
+                effective_batch_size = self.batch_size * self.grpo_samples  # = B x G
+                # torch.distributed.breakpoint()
+                grpo_stats: list[GRPOStats] = []
+                for _ in range(self._ppo_epochs):
+                    # batch_idxs = torch.randperm(effective_batch_size, device=self._device)
+                    batch_idxs = torch.arange(effective_batch_size, device=self._device)
+                    for i in range(0, effective_batch_size, self._ppo_batch_size):
+                        mini_batch_idxs = batch_idxs[i : i + self._ppo_batch_size]
+                        batch_grpo_stats: list[GRPOStats] = []
+
+                        for j in range(
+                            0, self._ppo_batch_size, self._ppo_backward_batch_size
+                        ):
+                            backward_batch_idxs = mini_batch_idxs[
+                                j : j + self._ppo_backward_batch_size
+                            ]
+
+                            batch_trajectory = GRPOTrajectory(
+                                *map(
+                                    partial(
+                                        torch.index_select,
+                                        dim=0,
+                                        index=backward_batch_idxs,
+                                    ),
+                                    trajectory,
+                                )
+                            )
+                            batch_grpo_stats.append(
+                                self._grpo_step(  # .backward() happens here
```

Comment:
nit, we don't need to make these private
```suggestion
                                self.grpo_step(  # .backward() happens here
```